keps/sig-apps/3189-downscale-pod-picker/README.md

```diff
@@ -0,0 +1,858 @@
+|---------------|------------------|------------------------------------------------------------------------------------------------|
+| `host`        | string           | Host name to connect to.                                                                       |
+| `httpHeaders` | HTTPHeader array | Custom headers to set in the request.                                                          |
+| `path`        | string           | Path to access on the HTTP server.                                                             |
+| `port`        | string / int     | Name or number of the port to access on the container. Number must be in the range 1 to 65535. |
+| `scheme`      | string           | Scheme to use for connecting to the host: `HTTP`, `HTTPS`. (default: `"HTTP"`)                 |
+
+The `httpHeaders` field has type [`HTTPHeader array`](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.23/#httpheader-v1-core)
+with the following specification:
+
+(NOTE: We will need to evaluate how feasible it is to extend `HTTPHeader v1 core` with `valueFrom` to enable mounting headers from Secrets,
+for authentication. This may need to be its own independent KEP.)
+
+| Field       | Type             | Description                                                            |
+|-------------|------------------|------------------------------------------------------------------------|
+| `name`      | string           | The header field name                                                  |
+| `value`     | string           | The header field value                                                 |
+| `valueFrom` | HTTPHeaderSource | Source for the header's value. Cannot be used if `value` is non-empty. |
+
+The `valueFrom` field has type `HTTPHeaderSource` with the following specification:
+
+| Field          | Type              | Description                                             |
+|----------------|-------------------|---------------------------------------------------------|
+| `secretKeyRef` | SecretKeySelector | Selects a key of a secret in the ReplicaSet's namespace |
+
+The `secretKeyRef` field has type [`SecretKeySelector`](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.23/#secretkeyselector-v1-core):
+
+| Field  | Type   | Description                                                       |
+|--------|--------|-------------------------------------------------------------------|
+| `key`  | string | The key of the secret to select from. Must be a valid secret key. |
+| `name` | string | Name of the referent.                                             |
+| `key`  | string | Specify whether the Secret or its key must be defined.            |
+
+
+#### "Pod Picker" API contract:
+
+The contract for a `downscalePodPicker` API will be as follows:
+
+- Request payloads to the API will contain:
```

Comment:
what HTTP method?  GET?  POST?  PUT?  PATCH?

Reply:
For HTTP, it pretty clearly should be POST.

However, as listed under goals, the API will be allowed to be gRPC (not just HTTP) in future releases, this is why I did not specify a protocol-specific feature in the API contract.

Reply:
"pretty clearly" doesn't fly in a spec :)

Reply:
Clarified in https://github.com/kubernetes/enhancements/pull/3190/commits/5086ff87ef00561292a2b0d612e3f06a7072451f that the HTTP implementation will use JSON encoding and POST requests in